source/keymap.js

```diff
@@ -11,8 +11,327 @@
+			} else {
+				this._pushBackHistory(ctx, fn);
+			}
+		},
+
+		/**
+		* When we press back key of R/C, window.history should be back 1 step.
+		* With synchronizing window.history and _backHistoryStack
+		* we do not want to run back key handler.
+		*
+		* @public
+		*/
+		ignorePopState: function() {
+			if (this.enableBackHistoryAPI) {
+				this._ignorePopState = true;
+				this._isPopStateInProgress = true;
+				history.go(-1);
+			}
+		},
+
+		/**
+		* @private
+		*/
+		callBackKeyHandler: function() {
+			this.isBackInProgress = true;
+			if (this._currentObj && this._handler) {
+				var fn = this._handler;
+				if (typeof fn == 'function') {
+					fn = fn.bind(this._currentObj);
+				}
+				fn();
+			}
+		},
+
+		/**
+		* @private
+		*/
+		popBackHistory: function() {
+			var bStack = this._backHistoryStack;
+			bStack.pop();
+			if (bStack.length) {
+				this._currentObj = bStack[bStack.length - 1].currentObj;
+				this._handler = bStack[bStack.length - 1].handler;
+			} else {
+				this._currentObj = this._handler = null;
+			}
+
+			this.isBackInProgress = false;
+		},
+
+		/**
+		* Dequeue _pushBackQueue
+		*
+		* @privae
+		*/
+		_dePushBackQueue: function() {
+			var queue = this._pushBackQueue,
+				length = queue.length,
+				item;
+			for (var i = 0; i < length; i++) {
+				item = queue.pop();
+				this._pushBackHistory(item.currentObj, item.handler);
+			}
+		},
+
+		/**
+		* Decide whether this popstate event should call handler or shouldn't.
+		*
+		* @public
+		*/
+		popStateHandler: function() {
+			this._isPopStateInProgress = false;
+			// Todo: We cannot prevent popstate event triggerd from history.go() or history.forward()
+			// If user call those event directly, moonstone controls may have unexpected behavior.
+			var state = !this._currentObj ? "empty"
+						: (this._ignorePopState || scope.ignoreFirstPopupEvent) ? "silence"
+						: !this._currentObj.getShowing() ? "invisible"
+						: "active";
+
+			switch (state) {
+			case 'empty':
+				break;
+			case 'silence':
+			//Popstate event should be ignored on following 2 conditions.
+			//1. When App is loaded, onpopstate event fired with null state.
+			//2. history.go(-1) triggers onpopstate event but it should be ignored.
+				scope.ignoreFirstPopupEvent = false;
+				this._ignorePopState = false;
+				this.doPopBackHistory();
+				break;
+			case 'invisible':
+			//Current back key target is on history and have handler too.
+			//However it is invisible.
+			//At this point, we should skip calling back key hanlder and restore history.
+				history.pushState({currentObjId: this._backHistoryStack[this._backHistoryStack.length - 1].currentObj.id}, '', '');
+				break;
+			case 'active':
+				this.callBackKeyHandler();
+				this.doPopBackHistory();
+				break;
+			}
+
+			this._dePushBackQueue();
+			return;
+		},
+
+		/**
+		* @private
+		*/
+		backKeyHandler: function (inSender, inEvent) {
+			switch (inEvent.keySymbol) {
```

Comment:
Do we anticipate handling more than one `keySymbol` value here? If not, it seems that we could remove the switch and simplify to a single conditional such as

```
if (inEvent.keySymbol == 'back' && this._currentObj && this._currentObj.getShowing())
```